[2단계 - 출석 다시 구현하기] 루키(김준하) 미션 제출합니다.

junhaa opened the pull request:
<!-- 




## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

안녕하세요 매트 😃
이번 step2에서는 출석부 미션을 처음부터 TDD로 구현했어요. 
step2 미션도 잘부탁드려요!

현재 코드에서는 2024년 12월을 기준으로 애플리케이션이 실행되고 있는데요. 현재 시간을 기준으로 변경하시려면 `Application` 에서 다음 코드의 주석을 제거해주세요 !!

```
//        LocalDateProvider localDateProvider = new SystemLocalDateProvider();
        LocalDateProvider localDateProvider = new FixedLocalDateProvider(LocalDate.of(2024, 12, 31));
```

## 체크 리스트

- [X] 미션의 필수 요구사항을 모두 구현했나요?
- [X] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [X] 애플리케이션이 정상적으로 실행되나요?
- [X] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [X] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

- 모든 원시값과 문자열을 포장하지 못했습니다.
- 3개 이상의 인스턴스 변수를 가진 클래스를 사용했습니다.
- 한 줄에 여러 개의 점을 사용하였습니다.
<!-- 이유 작성 -->

  
## 어떤 부분에 집중하여 리뷰해야 할까요?

<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->

## 1️⃣ 테스트를 위한 `LocalDate.now()`

이번 미션에서는 시스템 시간에 맞춰서 출석부 애플리케이션이 실행되도록 구현했어요. 그렇게 구현을 하다보니 테스트 코드에서는 시스템 시간에 맞춰서 테스트를 하면 일관되지 못한 테스트 결과를 나타낼 것이라고 생각했어요. 일관된 테스트 코드를 작성하기 위해, `LocalDate.now()`에 해당하는 부분을 추상화하여 테스트 코드에서는 원하는 시간을 반환할 수 있도록 변경했어요. 

하지만 이번 미션 내부에서 `LocalTime.now()`를 사용하는 클래스가 매우 많은데요, 현재는 이 클래스들이 생성자로 주입받고 있는 형식이라 클래스들이 너무 많은 인스턴스 변수로 가지는 것 같아요. 전역적으로 관리할 수 있는 방법이 있을까요..?! 전체적으로 추상화한 클래스에 대한 객체를 인스턴스로 저장하다 보니 너무 많은 인스턴스 변수를 가지는 것 같아요😭

## 2️⃣ `AttendanceManager`의 책임

이번 미션을 진행하면서 `Attendances`와 `Attendance` 객체 자체는 데이터를 저장하는 기능이라고 생각해서 내부 기능을 다른 클래스로 분리하였습니다. 그리고 전체적으로 `AttendanceManager` 클래스가 데이터를 관리하는 기능을 담당하게 되었는데, 이 클래스가 너무 많은 기능을 담당하는 건 아닌지 궁금합니다.

junhaa committed:
del: step2 미션을 위한 기존 코드 삭제

junhaa committed:
docs(README): 요구사항 및 기능 목록 정리

junhaa committed:
feat(AttendanceStatus): 출석 상태를 반환하는 기능 구현

junhaa committed:
feat(Attendance): 출석 시간과 출석 상태를 저장하는 클래스 구현

junhaa committed:
feat(Holiday): 공휴일 정보를 판별하는 기능 구현

junhaa committed:
feat: 캠퍼스의 운영 여부를 검증하는 기능 구현

junhaa committed:
feat: 출석 목록이 크루원의 이름에 해당하는지 검증하는 기능 구현

junhaa committed:
feat: 해당 날짜에 출석 기록이 있는 경우 예외가 발생하는 기능 구현

junhaa committed:
feat: 출석 수정 기능 구현

- 날짜, 등교시간을 통해 출석 내역 수정
- 수정하려는 날짜가 없는 경우 새로운 출석 기록 추가

junhaa committed:
feat: 출석 조회 기능 구현

- 전날까지의 출석, 지각, 결석 횟수를 반환하는 기능 구현

junhaa committed:
feat: 출석 정보 조회 시 정보가 없으면 결석으로 적용하는 기능 구현

junhaa committed:
feat: 출석 상태에 따라 위험 정도를 반환하는 기능 구현

junhaa committed:
refactor: 테스트 코드 역할에 맞게 위치 변경

- 캠퍼스 운영 일수를 반환하는 기능 이동

junhaa committed:
feat: 제적 위험 레벨을 계산하는 기능 구현

junhaa committed:
refactor: 목적에 맞게 클래스 이름 변경

- AttendanceBook -> AttendanceRecord로 이름 변경

junhaa committed:
refactor: Attendance를 저장하는 Map 일급 컬렉션으로 분리

junhaa committed:
refactor: 생성자 파라미터 순서 변경

- 시간, 날짜 순서에서 날짜, 시간 순서로 이해하기 쉽도록 변경

junhaa committed:
fix: 날짜 범위 수정

- IntStream에서 당일은 적용되지 않도록 수정

junhaa committed:
refactor: Attendances 내부 메서드 분리

- Attendances에서 출석 지각 결석 횟수를 계산하는 기능을 분리

junhaa committed:
refactor: Attendances 출석 정보 저장 시 Attendance를 생성하도록 변경

junhaa committed:
refactor: 클래스 이름 및 역할 변경

- 전체 크루의 출석 정보를 관리하는 역할에 맞게 AttendanceRecord -> AttendanceManager로 변경
- 크루 이름 별 출석 정보를 저장하여 인스턴스 변수로 저장

junhaa committed:
test: TestFixture 추가

- AttendanceManager 객체를 생성하여 반환하는 TestFixture 추가

junhaa committed:
refactor: 메서드 명 변경

junhaa committed:
feat: 특정 크루의 출석 정보를 반환하는 기능 구현

junhaa committed:
feat: 프로그램 실행 시 파일 데이터를 불러오는 기능 구현

junhaa committed:
feat: 사용자로부터 실행할 기능 번호를 입력받는 기능 구현

junhaa committed:
feat: 크루원의 닉네임을 입력받는 기능 구현

junhaa committed:
refactor: 패키지 구조 변경

- 클래스 기능에 맞게 패키지 이동

junhaa committed:
feat: 등교 시간을 입력 받는 기능 구현

- 출석 후 출석 기록을 출력

junhaa committed:
refactor: 출석 수정 시 불변 객체를 반환하도록 변경

- Attendance -> LocalTime 객체를 반환하도록 수정

junhaa committed:
feat: 출석 수정 시 입출력 기능 구현

- 수정하려는 날짜를 입력하는 기능
- 수정하려는 시간을 입력하는 기능

junhaa committed:
feat: 크루별 출석 기록을 출력하는 기능 구현

junhaa committed:
fix: 모든 기간의 데이터가 출력되는 오류 해결

- 해당하는 월의 데이터만 적용하도록 변경

junhaa committed:
style: 코드 컨벤션 적용

junhaa committed:
feat: 제적 위험자 목록 출력 기능 구현

junhaa committed:
refactor: LocalDate를 추상화하여 테스트 가능하도록 분리

junhaa committed:
fix: 출력 포맷 조정 밋 메시지 개선

junhaa committed:
test: AttendanceManager 관련 테스트 추가

junhaa committed:
test: DataLoader 테스트 추가

junhaa committed:
refactor: csv 파일 데이터 로드 위치 변경

- Controller 생성 시 -> Application 실행 시로 변경

hyeonic commented on src/main/java/attendance/domain/AttendanceManager.java:
```diff
@@ -0,0 +1,79 @@
+        Attendance prevAttendance = attendances.getCurrentAttendance(date);
+
+        attendances.addAttendance(date, time);
+        return Optional.ofNullable(prevAttendance)
+                .map(Attendance::time)
+                .orElse(null);
```
null을 else로 활용한다면 Optional 활용이 큰 의미가 없어보이는데 prevAttendance를 감싼 이유가 있을까요??

junhaa replied:
사실 기존에는 `Optional`을 사용하지 않고 바로 `Attendance`객체를 return 하는 형식을 생각했었어요. 

```java
    public Attendance modify(String crewName, LocalDate date, LocalTime time) {
        AttendanceChecker.checkCampusOpen(date, time);
        Attendances attendances = findAttendancesByName(crewName);
        Attendance prevAttendance = attendances.getCurrentAttendance(date);

        attendances.addAttendance(date, time);
        return prevAttendance;
    }
```

`Attendance`도 물론 불변성을 보장하지만, 추후 변경이 일어날 것을 생각하면 `LocalTime`을 넘기는 것이 좋지 않을까? 라는 생각이 들었어요. 그렇게 변경을 하다보니, LocalTime을 반환하는 부분에서 Attendance 객체 자체가 `null`인 경우 NPE이 발생해서 `Optional`로 반환하는 방식을 사용했어요.

```java
    public LocalTime modify(String crewName, LocalDate date, LocalTime time) {
        AttendanceChecker.checkCampusOpen(date, time);
        Attendances attendances = findAttendancesByName(crewName);
        Attendance prevAttendance = attendances.getCurrentAttendance(date);

        attendances.addAttendance(date, time);
        return prevAttendance.time(); // <-- NPE가 발생하는 부분
    }
```

junhaa replied:
위 내용은 아니지만, 비슷한 부분에서 제가 고민했던 부분에 대해서 궁금했던 점이 있어요. 🤔

우선 저번 1단계에서는 출석 기록이 없는 경우(등교를 안 한 경우), 새로운 `Attendance` 객체를 생성하고 내부에 **없는 출석 기록**을 의미하는 매직 넘버를 사용해서 저장했어요. 하지만 이번 2단계에서는 출석이 없는 경우 `null`값을 사용하는 것이 의미에 맞다고 생각해서 `null`값을 반환하도록 구현했어요. 하지만 `null`값을 반환하면 NPE가 발생할 수도 있고, `null`을 체크하는 코드가 증가할 수도 있다고 생각해요. 매트가 생각하기에는 `null`을 반환하는 것과 객체를 생성해 반환하는 것에 대해 어떻게 생각하시는 지 궁금해요!

hyeonic replied:
말씀 주신 것 처럼 null을 반환하는 것은 런타임에 돌아다닐 우려가 있기 때문에 `NullPointException`이 발생할 여지가 있을 것 같아요. 그렇기 때문에 null을 반환하는 것은 최대한 지양해야 할 것 같아요.

이런 경우 루키가 말씀 주신 것 null을 최대한 지양하기 위해 매직 넘버를 활용하거나 Null과 유사하게 사용할 수 있는 [Null Object 패턴](https://johngrib.github.io/wiki/pattern/null-object/)을 활용하거나 `getCurrentAttendance`에서 조회할 때 `Optional 타입`자체를 반환할 수도 있습니다. Optional 객체는 T 타입(여기에서는 Attendance가 될 수 있겠네요)의 객체가 있을 수도 없을 수도 있다는 것을 표현하기 위한 컨테이너 타입(어떠한 객체를 감싸는 타입이라고 생각해주세요)으로 활용될 수 있어요. 즉 Optional을 반환하게 되면 그것을 사용하는 쪽에게 `orElse`, `orElseGet`, `orElseThrow` 등을 활용하도록 하여 `null에 대한 처리를 위임`할 수도 있습니다. 

문제를 해결하는 방법에는 다양한 관점이 존재한다고 생각해요. 루키가 생각 했을 때 현재 구조에서 어떠한 방법이 가장 적합한지 잘 고민해보시고 선택하면 좋을 것 같아요 ㅎㅎ

junhaa replied:
Null Object 패턴이라는 게 따로 있는 줄 몰랐네요😮 링크해 주신 내용 읽고 나니 오히려 null 값에 대한 객체를 생성하는 게 null 검사 로직을 줄이고, 코드를 단순하게 작성할 수 있겠어요. 우선 Optional 타입 자체를 반환하는 형식으로 변경해보고, Null Object 패턴을 조금 더 공부해보고 null값을 사용해야 하는 상황을 마주한다면 도입해봐야겠어요. 사실 null값에 대해 고민이 많았었는데, 매트가 답변해 주신 내용을 보니 더욱 명확해 진 것 같아요. 답변 감사해요 매트 😄

hyeonic commented on src/main/java/attendance/domain/AttendanceManager.java:
```diff
@@ -0,0 +1,79 @@
+        this.statistics = statistics;
+    }
+
+    public void validateExistCrew(String crewName) {
+        if (!crewAttendances.containsKey(crewName)) {
+            throw new IllegalArgumentException("존재하지 않는 크루 닉네임입니다.");
+        }
+    }
```
저는 조건에 긍정문을 두는 것을 선호하는데요! 기회가 된다면 클린 코드 17장 냄새와 휴리스틱 쪽 읽어보면 좋을 것 같아요 ㅎㅎ

junhaa replied:
조건에 부정문을 넣으니 코드를 읽는 입장에서는 한번 더 생각해야 한다는 점이 있겠네요🤔 클린 코드 내용 참고해서 읽어보겠습니다! 감사합니다 😁

hyeonic commented on src/main/java/attendance/domain/AttendanceManager.java:
```diff
@@ -0,0 +1,79 @@
+import java.util.Map;
+import java.util.Optional;
+
+public class AttendanceManager {
```
> 이번 미션을 진행하면서 Attendances와 Attendance 객체 자체는 데이터를 저장하는 기능이라고 생각해서 내부 기능을 다른 클래스로 분리하였습니다. 그리고 전체적으로 AttendanceManager 클래스가 데이터를 관리하는 기능을 담당하게 되었는데, 이 클래스가 너무 많은 기능을 담당하는 건 아닌지 궁금합니다.

각각의 책임을 별도 객체에 두었다 하더라도 결국 여러 객체를 상호작용할 AttendanceManager와 같은 객체는 필연적으로 사용하게 되는 것 같아요. 그렇기 때문에 큰 문제는 아니라고 생각해요. 좀 더 개선할 여지가 있다면 getter를 지양하고 최대한 메시지를 던지는 방향으로 로직을 각각의 객체로 위임하거나 객체 분리를 통해 나눌 수 있을 것 같네요! 현재 구조에서는 크게 손 볼 곳은 없어보여서 괜찮아보여요 ㅎㅎ

hyeonic commented on src/main/java/attendance/domain/AttendanceStatistics.java:
```diff
@@ -0,0 +1,7 @@
+
+import java.util.Map;
+
+public interface AttendanceStatistics {
+    Map<AttendanceStatus, Integer> getTotalStatusCount(Attendances attendances);
+}
```
왜 별도 인터페이스를 구성하였을까요??

junhaa replied:
이번 step2에서는 `Attendances`가 `Attendance`를 관리하는 로직만 두고, 다른 로직들은 따로 static 클래스로 분리했어요. [관련 커밋](https://github.com/woowacourse/java-attendance/pull/133/commits/fd609a6bb99371677e959917192592237cc41ca5)

이후 static 메서드들은 오버라이딩이 불가능 하다는 점 때문에 별도의 인터페이스로 분리하여 추상화 하였습니다!

hyeonic replied:
좋습니다! 사용 방법이 익숙해지면 루키만의 추상화 기준도 잘 만들어지면 좋겠네요 ㅎㅎ

hyeonic commented on src/main/java/attendance/domain/LocalDateProvider.java:
```diff
@@ -0,0 +1,7 @@
+
+import java.time.LocalDate;
+
+public interface LocalDateProvider {
+    LocalDate now();
+}
```
> 이번 미션에서는 시스템 시간에 맞춰서 출석부 애플리케이션이 실행되도록 구현했어요. 그렇게 구현을 하다보니 테스트 코드에서는 시스템 시간에 맞춰서 테스트를 하면 일관되지 못한 테스트 결과를 나타낼 것이라고 생각했어요. 일관된 테스트 코드를 작성하기 위해, LocalDate.now()에 해당하는 부분을 추상화하여 테스트 코드에서는 원하는 시간을 반환할 수 있도록 변경했어요.
> 
> 하지만 이번 미션 내부에서 LocalTime.now()를 사용하는 클래스가 매우 많은데요, 현재는 이 클래스들이 생성자로 주입받고 있는 형식이라 클래스들이 너무 많은 인스턴스 변수로 가지는 것 같아요. 전역적으로 관리할 수 있는 방법이 있을까요..?! 전체적으로 추상화한 클래스에 대한 객체를 인스턴스로 저장하다 보니 너무 많은 인스턴스 변수를 가지는 것 같아요😭

인터페이스를 통한 생성자 주입으로 `유연함`은 가질 수 있었지만 객체의 관리 측면에서 고민을 하고 계시는군요 ㅎㅎ 하지만 static을 활용한 전역적인 관리는 도메인 내에 의존성을 전반적으로 가지기 때문에 변경에 취약할 수 있을 것 같아요.

레벨 1에서 다룰 내용은 아니지만 관점에 따라 객체를 생성하여 조합하는 책임을 별도로 다룰 수도 있을 것 같아요. 이건 루키가 불편함과 의문을 가지고 있다가 추후 레벨 2를 진행하며 자연스럽게 습득할 수 있을거라 생각해요. 순수한 java를 활용하여 애플리케이션을 개발 했을 때 어떠한 한계를 가지고 있는지 온전히 느껴보는 것도 좋아보입니다 ㅎㅎ

hyeonic commented on src/main/java/attendance/domain/DefaultAttendanceStatistics.java:
```diff
@@ -0,0 +1,58 @@
+        this.dateProvider = dateProvider;
+    }
+
+    @Override
+    public Map<AttendanceStatus, Integer> getTotalStatusCount(Attendances attendances) {
+        Map<AttendanceStatus, Integer> statusCount = new EnumMap<>(AttendanceStatus.class);
+        Map<LocalDate, Attendance> attendanceMap = attendances.getAttendances();
+        List<Attendance> attendanceList = extractValidAttendances(attendanceMap);
+        Arrays.stream(AttendanceStatus.values())
+                .forEach(status -> statusCount.put(status, getStatusCount(status, attendanceList)));
+
+        statusCount.put(ABSENCE, statusCount.get(ABSENCE) + countBlankAttendance(attendanceMap));
+        return statusCount;
+    }
```
변수에 자료형(List 등)을 표현하는 것은 너무 구체적인 정보를 표현하는 것 같아요!

junhaa replied:
변수명 작성하는 게 쉽지 않네요 😭 수정하겠습니다!

hyeonic commented on src/main/java/attendance/domain/Holiday.java:
```diff
@@ -0,0 +1,37 @@
+import java.time.LocalDate;
+import java.util.Arrays;
+
+public enum Holiday {
+    NEW_YEAR(1, 1),
+    INDEPENDENCE_MOVEMENT_DAY(3, 1),
+    CHILDRENS_DAY(5, 5),
+    MEMORIAL_DAY(6, 6),
+    LIBERATION_DAY(8, 15),
+    NATIONAL_FOUNDATION_DAY(10, 3),
+    HANGUL_DAY(10, 9),
+    CHRISTMAS(12, 25);
+
+    private final int month;
+    private final int day;
+
+    Holiday(int month, int day) {
+        this.month = month;
+        this.day = day;
+    }
+
+    public static boolean isHoliday(LocalDate date) {
+        return Arrays.stream(values())
+                .map(holiday -> LocalDate.of(date.getYear(), holiday.month, holiday.day))
+                .anyMatch(holidayDate -> holidayDate.equals(date));
+    }
+
+    public int getMonth() {
+        return month;
+    }
+
+    public int getDay() {
+        return day;
+    }
+}
```
휴일 관리를 enum을 통해 하셨군요 👍 이건 미션에 중요한 내용은 아니지만 양력이 아닌 음력을 기반한 휴일의 경우 매년 달라질 수 있기 때문에 enum을 통해 관리할 경우 매년 수정해줘야 할 수도 있다고 생각해요. enum을 사용 했을 때 내부 인스턴스 들이 자주 수정되어야 할 경우 어떻게 대응할 수 있는지, 이런 상황에서 enum이 적합한지에 대해서도 고민해보시죠 ㅎㅎ

junhaa replied:
Enum은 고정된 값이라 수정되지 않는 값을 사용해야 의미가 있다고 생각해요. 만약 양력이 아닌 음력으로 변경되는 경우 Enum 내부의 값들이 변경되어야 하기 때문에 적절하지 않은 것 같아요. 휴일에 대한 정보를 다른 방법으로 관리하도록 변경해보겠습니다! 🤔

hyeonic commented on src/main/java/attendance/domain/AttendanceChecker.java:
```diff
@@ -1,10 +1,55 @@
+        checkDate(date);
+        checkTime(time);
+    }
+
+    public static boolean isCampusOpenDate(LocalDate date) {
+        return !isWeekend(date) && !Holiday.isHoliday(date);
+    }
+
+    private static void checkDate(LocalDate date) {
+        if (!isCampusOpenDate(date)) {
+            throw new IllegalArgumentException(String.format(
+                    CAMPUS_CLOSED_ERROR_FORMAT,
+                    date.getMonthValue(),
+                    date.getDayOfMonth(),
+                    getDisplayName(date)
+            ));
+        }
+    }
+
+    private static boolean isWeekend(LocalDate date) {
+        DayOfWeek dayOfWeek = date.getDayOfWeek();
+        return dayOfWeek == DayOfWeek.SATURDAY || dayOfWeek == DayOfWeek.SUNDAY;
+    }
+
+    private static String getDisplayName(LocalDate date) {
+        return date.getDayOfWeek().getDisplayName(TextStyle.FULL, Locale.KOREA);
+    }
+
+    private static void checkTime(LocalTime time) {
+        LocalTime openTime = LocalTime.of(CAMPUS_OPEN_HOUR, CAMPUS_OPEN_MINUTE);
+        LocalTime closeTime = LocalTime.of(CAMPUS_CLOSE_HOUR, CAMPUS_CLOSE_MINUTE);
+
+        if (time.isAfter(closeTime) || time.isBefore(openTime)) {
+            throw new IllegalArgumentException("현재 캠퍼스 운영시간이 아닙니다.");
+        }
+    }
 
-    boolean isCampusDay(int day);
 
 }
```
AttendanceChecker 내에 명시한 내용을 살펴보면 때에 따라 Attendance, Attendances 등으로 도메인 내로 위치 시킬 수 있지 않을까 생각이 되어요. static을 활용했을 때 얻을 수 있는 장단점에 대해 고민해보면 좋을 것 같아요.

junhaa replied:
이 부분은 질문했던 문제점들 때문에 변경하지 못한 부분이 있어요.

> 하지만 이번 미션 내부에서 LocalTime.now()를 사용하는 클래스가 매우 많은데요, 현재는 이 클래스들이 생성자로 주입받고 있는 형식이라 클래스들이 너무 많은 인스턴스 변수로 가지는 것 같아요. 전역적으로 관리할 수 있는 방법이 있을까요..?! 전체적으로 추상화한 클래스에 대한 객체를 인스턴스로 저장하다 보니 너무 많은 인스턴스 변수를 가지는 것 같아요😭

이 내용과 비슷한 부분인데, AttendanceChecker 클래스를 추상화하여 사용하려고 했어요. 하지만 AttendanceChecker 클래스를 추상화하게 되면 너무 많은 클래스에서 변수로 가지게 되어 결국 수정하지 않았어요. 매트가 답변해주신 것 처럼 static을 사용하는 경우에는 도메인에서 전반적으로 의존성을 가지기 때문에 추상화하는 방법이 더 좋을 것 같아요. 수정해보겠습니다!

hyeonic reviewed (changes requested):
안녕하세요 루키~ 커밋 기록을 살펴보았는데 최대한 작은 단위로 TDD를 시도하려는 모습이 보였어요 ㅎㅎ

간단히 코멘트 남겼습니다! 코드 변경보다 현재 구조에 대한 루키의 생각을 물어보는 것 위주라 고민해보시고 남겨주세요 😃

junhaa committed:
refactor: AttendanceChecker 추상화

- static 메서드 추상화를 통해 수정가능하도록 변경

junhaa committed:
refactor: 존재 여부 검증 로직을 긍정문으로 변경하여 가독성 개선

junhaa committed:
refactor: 변수명에 자료형을 제거하여 가독성 개선

junhaa committed:
refactor: 공휴일에 대한 정보 CSV파일을 통해 가져오도록 수정

- Enum의 형식 대신 AttendanceChecker에서 공휴일 여부를 판단하도록 수정
- 공휴일에 대한 정보는 수정 가능하도록 CSV 파일로 가져오도록 변경

junhaa commented:
안녕하세요 매트 😃

피드백 주신 내용을 읽고 수정해보았어요.

수정해야 할 부분이 있으면 피드백 주시면 바로 수정해보겠습니다!

감사합니다 😄

hyeonic commented on src/main/java/attendance/util/CsvDataLoader.java:
```diff
@@ -0,0 +1,72 @@
+import java.util.List;
+import java.util.Map;
+
+public class CsvDataLoader implements DataLoader{
```
휴일 관련 로직을 별도 인터페이스를 통해 구성하였군요 ㅎㅎ 👍 enum으로 구성한 이전과 달리 csv 파일을 통해 유연하게 관리할 수 있겠네요!

잘 작성하셨는데 간단히 컨벤션 확인 부탁드릴게요.

hyeonic commented on src/main/java/attendance/domain/Holiday.java:
```diff
@@ -0,0 +1,31 @@
+import java.time.LocalDate;
+import java.util.Arrays;
+
+public class Holiday {
+    private final String name;
+    private final int month;
+    private final int day;
+
+    public Holiday(String name, int month, int day) {
+        this.name = name;
+        this.month = month;
+        this.day = day;
+    }
+
+    public boolean isMatch(LocalDate date) {
+        if(date.getMonthValue() == month && date.getDayOfMonth() == day){
+            return true;
+        }
+        return false;
+    }
+
+    public int getMonth() {
+        return month;
+    }
+
+    public int getDay() {
+        return day;
+    }
+}
```
불필요한 필드 및 메서드는 제거해도 좋을 것 같아요. 추가로 컨벤션 확인 부탁드려요.

hyeonic reviewed (changes requested):
안녕하세요 루키! 대부분 잘 개선 된 것 같아요 ㅎㅎ 다만 새롭게 구성한 코드 쪽에 컨벤션이 맞지 않는 부분이 있어서 그 부분만 확인되면 바로 머지해도 될 것 같아요!

null과 Optional 관련해서 질문 주신 것은 아래 남겨 두었으니 확인 부탁드려요 😃

junhaa committed:
refactor: 불필요한 메서드 제거

- 사용하지 않는 getter 메서드 제거

junhaa committed:
style: 코드 컨벤션 적용

junhaa committed:
refactor: null값을 반환하지 않고 Optional로 반환하도록 변경

- null값 검증에 대한 처리를 위임하기 위해 Optional로 변경하여 반환

junhaa commented:
안녕하세요 매트 😃

피드백 주신 내용을 읽고 수정해보았어요.

부족한 점이 많은 코드인데, 상세하게 피드백 해주셔서 감사합니다😅

혹시라도 수정해야 할 부분이 있으면 수정하도록 하겠습니다 감사해요 매트 😄

hyeonic commented:
앗 Optional 객체는 반환 타입이 아닌 파라미터에서 사용하는 것을 지양하고 있어요 ㅠㅜ 결국 Optional 객체 자체가 null인지 아닌지를 추가적으로 검토해야 하기 때문입니다. 관련해서 아마 IntelliJ도 힌트를 주고 있을 거예요.

![image](https://github.com/user-attachments/assets/dd36c81a-04bf-416c-bd65-e5c42d9081d6)

Optional 객체를 그대로 전달하기 보단 반환 받은 즉시 소모하는 편이 좋아보이네요.

 * [Why should Java 8's Optional not be used in arguments](https://stackoverflow.com/questions/31922866/why-should-java-8s-optional-not-be-used-in-arguments)

junhaa committed:
refactor: LocalTime에서 null 값을 판별하기 위해 추상 클래스로 분리

junhaa committed:
refactor: LocalTime을 사용하는 객체를 NullableLocalTime의 구현체로 변경

junhaa commented:
앗 제가 Intellij 경고 표시를 잘 못봤네요 😅 Optional 객체를 파라미터로 넘기는 경우에는 오히려 Optional 객체를 판별해야 하기 때문에 내부 코드가 더욱 복잡해지는 것 같아요. 매트가 링크 해주신 내용을 읽고, 따로 객체를 만들어서 null 값을 판별하도록 변경했습니다! 감사해요 매트 😄

hyeonic reviewed (approved):
안녕하세요 루키~ 처음 해보는 TDD라 어려움이 많았을 것 같은데 잘 적용하신 것 같아서 이만 머지해도 좋을 것 같아요. 늦은 시간까지 리뷰 반영하느라 고생하셨습니다!

저희가 주고 받은 대화 속에서 루키만의 근거를 만들어가는데 도움이 되었길 바라봅니다. 다음 미션도 화이팅 👏

hyeonic commented:
참고로 Optional 객체를 `반환 타입`으로 사용하는 것은 종종 볼 수 있을거예요 ㅎㅎ 메서드의 파라미터로 전달될 때만 왜 지양하는지 잘 알아두면 좋을 것 같네요!
